### DISCUSSION TITLE
matching URL-as-parameter with catch-all placeholder

### DISCUSSION THREAD

**Discussion by guillomovitch** [mojolicious/mojo#1722] [Q&A]
Hi. I'm not sure this is the correct place for such kind of questions, just tell me if I should use another media.

I'm trying to allow matching business objects (SAML entities, actually) through business identifier (URN or URI), in addition to application-specific internal ID. So, my route dispatching looks like:
```
my $entities = $routes->any('/entities')->to(controller => 'Entities');
$entities->get('/view/<id:num>')->to(action => 'view')->name('view_entity');
$entities->get('/view/*id')->to(action => 'view')->name('view_entity_saml_id');
```

Using an URN (/entities/view/urn:foo:bar:tld) as identifier work OK, but using an URI (https://foo.bar.tld) doesn't:
- If I URL-encode the parameter (/entities/view/https%3A%2F%2Ffoo.bar.tld), no matching route is found
- If I doesn't URL-encode the parameter (/entities/view/https://foo.bar.tld), the controller receive a truncated value for the parameter, with double slashes sequences sequence reduced to a single one (https:/foo.bar.tld)

Is there a technical solution ? And does it make sense to coerce this kind of value into a path parameter, instead of a query parameter ?

Regards.
-- 
Guillaume

**Comment by Grinnz** [comment]
I would not recommend relying on slashes as data within the URL. A query parameter is a reasonable way to receive such a parameter.

**Reply by guillomovitch** [reply]
sure, if it was the general rule to use query parameters. However, this application already use path info for standard internal IDs, everywhere, such as, I'd rather avoid a different syntax for the few kind of objects that also have business IDs as alternative.

**Comment by zoffixznet** [comment]
> If I URL-encode the parameter (/entities/view/https%3A%2F%2Ffoo.bar.tld), no matching route is found

Are you using Apache by any chance? You [need to set `AllowEncodedSlashes On`](https://httpd.apache.org/docs/2.4/mod/core.html#allowencodedslashes)

This test app works fine, when run on its own, via `morbo` without anything in front of it:
```perl
use Mojolicious::Lite;
get '/view/*id' => sub { my $c = shift; $c->render(text => 'I ♥ Mojolicious! ID: ' . $c->param("id")) };
app->start;
```

**Reply by guillomovitch** [reply]
Thanks for the tip, it was enough to fix the issue :+1:
